onyo/lib/onyo.py

```diff
@@ -327,20 +346,29 @@ def _get_files_untracked(self) -> set[Path]:
         untracked = {Path(x) for x in self._git(['ls-files', '-z', '--others', '--exclude-standard']).split('\0') if x}
         return untracked
 
-    def _get_root(self) -> Path:
+    @staticmethod
+    def _find_root(directory: Path) -> Path:
         """
+        Find and return the root of an Onyo repository (containing `.git/` and
+        `.onyo/`) for a given `Path` inside of an existing repository.
+
+        If the `directory` is not inside an existing repository, an error is
+        raised.
         """
+        root = None
+
         try:
-            root = self._git(['rev-parse', '--show-toplevel'], cwd=self._opdir).strip()
-        except subprocess.CalledProcessError:
-            log.error(f"'{self._opdir}' is not a Git repository.")
-            raise OnyoInvalidRepoError(f"'{self._opdir}' is not a Git repository.")
+            ret = subprocess.run(["git", "rev-parse", "--show-toplevel"],
+                                 cwd=directory, check=True,
+                                 capture_output=True, text=True)
+            root = Path(ret.stdout.strip()).resolve()
```

Comment:
Minor nitpick: L364 should be outside of the try/except block, because the only thing evaluated is the subprocess call.

Reply:
I don't think so. It's not only about what is possibly raising the error, it's also about what is executed as a whole if there was no error. With that line being in the block, it's only ever executed when `ret` actually should contain something that this line can work with.
If you pull it out, you have branch twice for the same error:

1. try .. except
2. `if ret ... else`, which needs to "reconstruct" whether or not an exception was raised.

Reply:
What you're saying is case-specific. It's true for some cases, but not this one. If a `subprocess.CalledProcessError` or `FileNotFoundError` is raised, an exception is raised and the code will never reach `root`. In all other cases `root` will behave the same as it does inside the block. In this case it is better to be clear about what is being evaluated.

Reply:
What you are saying is: "Yes, this should only be executed if the previous line succeeded, but I want this line to be written in a piece of code, that doesn't know that to be the case. Instead, I - the dev - know, that at this particular moment of the development, the `except` clause raises again, thus implicitly making sure this has the same effect."

Is that assessment wrong?

Reply:
I think that assessment is not addressing the full problem. Say we don't have that re-raise.
```python3
try:
    ret = some_code()
    root = ret.some_property
except SomeError:
   log.warning('something went wrong')

do_something(root)
```

In this case, things would crash on the last line if `some_code()` raises an exception. I would have to change things in the except block, so that we get:
```python3
try:
    ret = some_code()
    root = ret.some_property
except SomeError:
    root = None

do_something(root)
```

Now everything works fine. However, should we put `do_something(root)` also in the try/except block? It's using the contents of root, just like root is using the contents of ret, after all, and only I - the dev - know that in the except block we set it to `None` if an exception is raised. Note, though, that I think the second example is a valid use-case where `root` can be in the `try` block, because the `except` block specifically deals with it. In the current case, the try/except block doesn't deal with `root` at all (but placing it there gives the impression it does, which is the root (see what I did there 😆) of my problem).

Furthermore by this reasoning, `root = None` makes no sense, because `Path(None)` raises an exception. The only reason it's in there is to satisfy pyre, which doesn't understand that an exception is raised if `root` is not set.

Reply:
The problem with your version in this particular case for me is:

You have line `root = Path(ret.stdout.strip()).resolve()` whose validity entirely depends on the previous `try .. else` block, which it is not a part of. That's fine in this state of the code. The issue comes when other changes are introduced over time. Those to parts are semantically coupled, but not structurally. This can fall apart over time. At the same time: It does not provide any advantage over the existing solution. It can only make things worse in the long-run.